java/arcs/core/data/expression/Stringifier.kt

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+package arcs.core.data.expression
+
+/** Traverses a tree of [Expression] objects, converting each node to manifest format (String). */
+class ExpressionStringifier(val parameterScope: Expression.Scope = ParameterScope()) :
+    Expression.Visitor<String> {
+    /*
+     * TODO: implement a precedence visitor to optimize parenthesis removal
+     * If a parent expression has higher precedence than a child node, you don't need to emit
+     * parens. e.g. let multiply/div = precedence 2, add/sub = precedence 1, then the
+     * tree (* (+ 1 2) 3) representing 3 * (1 + 2) needs parens because precedence * > precedence +
+     */
+    private fun <E> maybeParen(expr: Expression<E>) = when (expr) {
+        !is Expression.BinaryExpression<*, *, *> -> expr.accept(this)
+        else -> "(" + expr.accept(this) + ")"
+    }
+
+    override fun <E, T> visit(expr: Expression.UnaryExpression<E, T>) =
+        expr.op.token + expr.expr.accept(this)
+
+    override fun <L, R, T> visit(expr: Expression.BinaryExpression<L, R, T>) =
+        maybeParen(expr.left) + " ${expr.op.token} " + maybeParen(expr.right)
+
+    override fun <E : Expression.Scope, T> visit(expr: Expression.FieldExpression<E, T>) =
+        expr.qualifier.accept(this) + ".${expr.field}"
```

Comment:
This helps me put `FieldExpression`s into context. This is a cool implementation for `.` operator access.